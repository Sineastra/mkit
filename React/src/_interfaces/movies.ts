export interface movieIntF {
	_id?: string,
	externalId: string,
	poster: string,
	title: string,
	genres: string[],
	runtime: number,
	officialSite: string,
	description: string,
}

export interface movieDetailsForUserIntF {
	rating: number
	notes: string
}