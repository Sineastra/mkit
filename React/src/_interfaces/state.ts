export interface movieInteractionIntF {
	_id: string,
	rating: number,
	privateComment: string,
}
export interface userData {
	_id: string,
	name: string,
	iat?: number,
}
export interface userSliceIntF {
	loggedUser: userData | null
}