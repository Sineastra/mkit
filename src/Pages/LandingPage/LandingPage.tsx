import HeroSection from "../../Components/HeroSection/HeroSection"


const LandingPage = () => {

	return (
		<>
			<HeroSection/>
		</>
	)
}

export default LandingPage