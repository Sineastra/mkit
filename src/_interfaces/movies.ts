export interface FavCardMovieIntF {
	imdbId: string,
	poster: string,
	title: string,
	genre: string,
}
