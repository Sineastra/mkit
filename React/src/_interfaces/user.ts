import { movieIntF } from "./movies"


export interface user {
	_id: string,
	name: String,
	hashedPassword: String,
	favourites: movieIntF[],
	notes: { movieId: string, note: string }[],
	ratings: { movieId: string, rating: string }[],
}
