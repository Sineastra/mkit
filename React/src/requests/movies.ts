import { FavCardMovieIntF, SearchCardMovieIntF } from "../_interfaces/movies"
import { baseAPIRequest } from "./baseRequest"


const getDetailedInfo = (movie: any): SearchCardMovieIntF => {
	return {
		imdbId: movie.externals.imdb,
		poster: movie.image.medium,
		title: movie.name,
		genres: movie.genres,
		runtime: movie.runtime,
		officialSite: movie.officialSite,
		description: movie.summary,
	}
}

interface moviesRequestsIntF {
	getFavs: (ids: string[]) => Promise<FavCardMovieIntF[]>,
	getSearchs: (query: string) => Promise<SearchCardMovieIntF[]>,
	getDetails: (id: string) => Promise<SearchCardMovieIntF>
}
const movieRequests: moviesRequestsIntF = {
	getFavs: async (ids) => {
		const fetchReq = async (id: string): Promise<FavCardMovieIntF> => {
			const movieData = await baseAPIRequest(`/lookup/shows?imdb=${ id }`)

			return {
				imdbId: movieData.externals.imdb,
				poster: movieData.image.medium,
				title: movieData.name,
				genres: movieData.genres,
			}
		}
		const promises = ids.map(fetchReq)
		const movies: FavCardMovieIntF[] = await Promise.all(promises)

		return movies
	},
	getSearchs: async (query) => {
		const movieData = await baseAPIRequest(`/search/shows?q=${ query }`)
		const movies: SearchCardMovieIntF[] = movieData.map((x: any) => {
			return getDetailedInfo(x)
		})

		return movies
	},
	getDetails: async (id) => {
		const movieData = await baseAPIRequest(`/lookup/shows?imdb=${ id }`)

		return getDetailedInfo(movieData)
	},
}

export { movieRequests }