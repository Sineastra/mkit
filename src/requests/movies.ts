import { FavCardMovieIntF, SearchCardMovieIntF } from "../_interfaces/movies"
import { baseAPIRequest } from "./baseRequest"


interface moviesRequestsIntF {
	getFavs: (ids: string[]) => Promise<FavCardMovieIntF[]>,
	getSearchs: (query: string) => Promise<SearchCardMovieIntF[]>,
}

const movieRequests: moviesRequestsIntF = {
	getFavs: async (ids) => {
		const fetchReq = async (id: string): Promise<FavCardMovieIntF> => {
			const movieData = await baseAPIRequest(`/lookup/shows?imdb=${ id }`)

			return {
				imdbId: movieData.externals.imdb,
				poster: movieData.image.medium,
				title: movieData.name,
				genres: movieData.genres,
			}
		}
		const promises = ids.map(fetchReq)
		const movies: FavCardMovieIntF[] = await Promise.all(promises)

		return movies
	},
	getSearchs: async (query) => {
		const movieData = await baseAPIRequest(`/search/shows?q=${ query }`)
		const movies: SearchCardMovieIntF[] = movieData.map((x: any) => {
			return {
				imdbId: x.externals.imdb,
				poster: x.image.medium,
				title: x.name,
				genres: x.genres,
				runtime: x.runtime,
				officialSite: x.officialSite,
				description: x.summary,
			}
		})

		return movies
	},
}

export { movieRequests }