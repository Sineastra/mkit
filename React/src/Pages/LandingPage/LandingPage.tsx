import HeroSection from "../../Components/HeroSection/HeroSection"
import Favourites from "../../Components/Favourites/Favourites"
import { FavCardMovieIntF } from "../../_interfaces/movies"
import { useSelector } from "react-redux"
import { RootState } from "../../_state/app/store"
import { useEffect, useState } from "react"
import { movieRequests } from "../../requests/movies"

const LandingPage = () => {
	const [movies, setMovies] = useState<FavCardMovieIntF[]>([])
	const moviesIds: string[] = useSelector((state: RootState) => state.userData.favourites)

	useEffect(() => {
		const getData = async () => {
			if (moviesIds.length > 0) {
				const data = await movieRequests.getFavs(moviesIds)

				setMovies(data)
			}
		}

		getData()
	}, [moviesIds])

	return (
		<>
			<HeroSection/>
			<Favourites movies={ movies }/>
		</>
	)
}

export default LandingPage