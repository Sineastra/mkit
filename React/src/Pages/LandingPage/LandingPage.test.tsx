import { configureStore } from "@reduxjs/toolkit"
import userReducer, { changeUser } from "../../_state/features/userSlice"
import renderer from "react-test-renderer"
import { Provider } from "react-redux"
import { BrowserRouter } from "react-router-dom"
import LandingPage from "./LandingPage"
import { movieIntF } from "../../_interfaces/movies"

const baseMockedMovie: movieIntF = {
	externalId: 'a',
	poster: 'b',
	title: 'b',
	genres: ['b'],
	runtime: 1,
	officialSite: 'b',
	description: 'b',
}
let mockedMovies: movieIntF[] = []

const store = configureStore({
	reducer: {
		userData: userReducer,
	},
})

jest.mock("../../requests/movies", () => {
	return {
		movieRequests: {
			getFavs: jest.fn(() => mockedMovies),
		},
	}
})

describe("---> Testing the Landing Page", () => {
	const renderTree = () => renderer.create(
		<Provider store={ store }>
			<BrowserRouter>
				<LandingPage/>
			</BrowserRouter>
		</Provider>,
	).toJSON()

	beforeAll(() => {
		store.dispatch(changeUser({ _id: '1', name: 'a' }))
	})

	beforeEach(() => {
		jest.clearAllMocks()
	})

	it("snapshot test -> renders correctly with no favourites", () => {
		const view = renderTree()

		expect(view).toMatchSnapshot()
	})
	it("snapshot test -> renders correctly WITH favourites", () => {
		mockedMovies = [baseMockedMovie]
		const view = renderTree()

		expect(view).toMatchSnapshot()
	})
})