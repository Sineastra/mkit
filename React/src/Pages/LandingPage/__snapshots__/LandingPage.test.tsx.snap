// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`---> Testing the Landing Page snapshot test -> renders correctly WITH favourites 1`] = `null`;

exports[`---> Testing the Landing Page snapshot test -> renders correctly with no favourites 1`] = `null`;
